<center><h1>Wang_Timothy_HW3</h1></center>
<br>
<br>

Name: Timothy Wang
<br>
Github Username: timothyswang
<br>
USC ID: 2697040790

## 1. Time Series Classification Part 1: Feature Creation/Extraction

### (a) Download Data

Package imports

In [1]:
import os
import pandas as pd
import numpy as np
from scipy.stats import bootstrap
# import bootstrapped.bootstrap as boot
# import bootstrapped.stats_functions as bootstrap_stat

# bootstrapped.bootstrap

Get the AReM Data Set

In [2]:
directory_bending1 = "../data/AReM/bending1"
directory_bending2 = "../data/AReM/bending2"
directory_cycling = "../data/AReM/cycling"
directory_lying = "../data/AReM/lying"
directory_sitting = "../data/AReM/sitting"
directory_standing = "../data/AReM/standing"
directory_walking = "../data/AReM/walking"

### (b) Test and Train Data

In [3]:
data_test = []
data_train = []
data_all = []

data_test_names = []
data_train_names = []

#[1]
for this_file in os.listdir(directory_bending1):
    path_this_file = directory_bending1 + "/" + this_file
    #[2]
    df_this_file = pd.read_csv(path_this_file)

    if this_file == "dataset1.csv" or this_file == "dataset2.csv":
        # print(this_file)
        data_test.append(df_this_file)
        data_test_names.append(path_this_file)
    else:
        data_train.append(df_this_file)
        # print(this_file)
        data_train_names.append(path_this_file)

    data_all.append(df_this_file)

#[1]
for this_file in os.listdir(directory_bending2):
    path_this_file = directory_bending2 + "/" + this_file

    #[2]
    df_this_file = pd.read_csv(path_this_file)

    if this_file == "dataset1.csv" or this_file == "dataset2.csv":
        # print(this_file)
        data_test.append(df_this_file)
        data_test_names.append(path_this_file)
    else:
        data_train.append(df_this_file)
        # print(this_file)
        data_train_names.append(path_this_file)

    data_all.append(df_this_file)

#[1]
for this_file in os.listdir(directory_cycling):
    path_this_file = directory_cycling + "/" + this_file

    #[2]
    df_this_file = pd.read_csv(path_this_file)

    if this_file == "dataset1.csv" or this_file == "dataset2.csv" or this_file == "dataset3.csv":
        # print(this_file)
        data_test.append(df_this_file)
        data_test_names.append(path_this_file)
    else:
        data_train.append(df_this_file)
        # print(this_file)
        data_train_names.append(path_this_file)

    data_all.append(df_this_file)

#[1]
for this_file in os.listdir(directory_lying):
    path_this_file = directory_lying + "/" + this_file

    #[2]
    df_this_file = pd.read_csv(path_this_file)

    if this_file == "dataset1.csv" or this_file == "dataset2.csv" or this_file == "dataset3.csv":
        # print(this_file)
        data_test.append(df_this_file)
        data_test_names.append(path_this_file)
    else:
        data_train.append(df_this_file)
        # print(this_file)
        data_train_names.append(path_this_file)

    data_all.append(df_this_file)

#[1]
for this_file in os.listdir(directory_sitting):
    path_this_file = directory_sitting + "/" + this_file

    #[2]
    df_this_file = pd.read_csv(path_this_file)

    if this_file == "dataset1.csv" or this_file == "dataset2.csv" or this_file == "dataset3.csv":
        # print(this_file)
        data_test.append(df_this_file)
        data_test_names.append(path_this_file)
    else:
        data_train.append(df_this_file)
        # print(this_file)
        data_train_names.append(path_this_file)

    data_all.append(df_this_file)

#[1]
for this_file in os.listdir(directory_standing):
    path_this_file = directory_standing + "/" + this_file

    #[2]
    df_this_file = pd.read_csv(path_this_file)

    if this_file == "dataset1.csv" or this_file == "dataset2.csv" or this_file == "dataset3.csv":
        # print(this_file)
        data_test.append(df_this_file)
        data_test_names.append(path_this_file)
    else:
        data_train.append(df_this_file)
        # print(this_file)
        data_train_names.append(path_this_file)

    data_all.append(df_this_file)

#[1]
for this_file in os.listdir(directory_walking):
    path_this_file = directory_walking + "/" + this_file

    #[2]
    df_this_file = pd.read_csv(path_this_file)

    if this_file == "dataset1.csv" or this_file == "dataset2.csv" or this_file == "dataset3.csv":
        # print(this_file)
        data_test.append(df_this_file)
        data_test_names.append(path_this_file)
    else:
        data_train.append(df_this_file)
        # print(this_file)
        data_train_names.append(path_this_file)

    data_all.append(df_this_file)

### (c) Feature Extraction

#### i. Research

Some forms of "primitive" time-domain features in time series classification includes mean, standard deviation, and slope [3] (as well as the other features that we'll look at in the next part, such as minimum, maximum, median, first quartile, and third quartile). Other features could be other series, such as a new series created using Fourier transforms, or fitted auto-regressive coefficients [3].

Other series type features can include a difference transform (where you take the difference in values between each consecutive datapoints) -- this is known as "first order difference" [4]. Another series type featurer is standardization, where you subtract the mean and divide by the standard deviation such that you get a Gaussian (normal) distribution [4]. You could have a power transform, such as taking the log of all of your values (only if your values are positive) [4]. Finally, you could use normalization, where you rescale your data range to be between 0 and 1 [4].

#### ii. Extraction

In [4]:
time_domain_features = []
instance = 1

# testing_df = data_test[0]
# print(testing_df.describe())

def get_time_features(this_df, test_or_train, instance):
    avg_rss12_min = this_df.describe()["avg_rss12"]["min"]
    avg_rss12_max = this_df.describe()["avg_rss12"]["max"]
    avg_rss12_mean = this_df.describe()["avg_rss12"]["mean"]
    avg_rss12_median = this_df.describe()["avg_rss12"]["50%"]
    avg_rss12_std = this_df.describe()["avg_rss12"]["std"]
    avg_rss12_firstq = this_df.describe()["avg_rss12"]["25%"]
    avg_rss12_thirdq = this_df.describe()["avg_rss12"]["75%"]
    avg_rss12_all = [avg_rss12_min, avg_rss12_max, avg_rss12_mean, avg_rss12_median, avg_rss12_std, avg_rss12_firstq, avg_rss12_thirdq]

    var_rss12_min = this_df.describe()["var_rss12"]["min"]
    var_rss12_max = this_df.describe()["var_rss12"]["max"]
    var_rss12_mean = this_df.describe()["var_rss12"]["mean"]
    var_rss12_median = this_df.describe()["var_rss12"]["50%"]
    var_rss12_std = this_df.describe()["var_rss12"]["std"]
    var_rss12_firstq = this_df.describe()["var_rss12"]["25%"]
    var_rss12_thirdq = this_df.describe()["var_rss12"]["75%"]
    var_rss12_all = [var_rss12_min, var_rss12_max, var_rss12_mean, var_rss12_median, var_rss12_std, var_rss12_firstq, var_rss12_thirdq]

    avg_rss13_max = this_df.describe()["avg_rss13"]["max"]
    avg_rss13_min = this_df.describe()["avg_rss13"]["min"]
    avg_rss13_mean = this_df.describe()["avg_rss13"]["mean"]
    avg_rss13_median = this_df.describe()["avg_rss13"]["50%"]
    avg_rss13_std = this_df.describe()["avg_rss13"]["std"]
    avg_rss13_firstq = this_df.describe()["avg_rss13"]["25%"]
    avg_rss13_thirdq = this_df.describe()["avg_rss13"]["75%"]
    avg_rss13_all = [avg_rss13_min, avg_rss13_max, avg_rss13_mean, avg_rss13_median, avg_rss13_std, avg_rss13_firstq, avg_rss13_thirdq]

    var_rss13_min = this_df.describe()["var_rss13"]["min"]
    var_rss13_max = this_df.describe()["var_rss13"]["max"]
    var_rss13_mean = this_df.describe()["var_rss13"]["mean"]
    var_rss13_median = this_df.describe()["var_rss13"]["50%"]
    var_rss13_std = this_df.describe()["var_rss13"]["std"]
    var_rss13_firstq = this_df.describe()["var_rss13"]["25%"]
    var_rss13_thirdq = this_df.describe()["var_rss13"]["75%"]
    var_rss13_all = [var_rss13_min, var_rss13_max, var_rss13_mean, var_rss13_median, var_rss13_std, var_rss13_firstq, var_rss13_thirdq]

    avg_rss23_max = this_df.describe()["avg_rss23"]["max"]
    avg_rss23_min = this_df.describe()["avg_rss23"]["min"]
    avg_rss23_mean = this_df.describe()["avg_rss23"]["mean"]
    avg_rss23_median = this_df.describe()["avg_rss23"]["50%"]
    avg_rss23_std = this_df.describe()["avg_rss23"]["std"]
    avg_rss23_firstq = this_df.describe()["avg_rss23"]["25%"]
    avg_rss23_thirdq = this_df.describe()["avg_rss23"]["75%"]
    avg_rss23_all = [avg_rss23_min, avg_rss23_max, avg_rss23_mean, avg_rss23_median, avg_rss23_std, avg_rss23_firstq, avg_rss23_thirdq]

    var_rss23_min = this_df.describe()["var_rss23"]["min"]
    var_rss23_max = this_df.describe()["var_rss23"]["max"]
    var_rss23_mean = this_df.describe()["var_rss23"]["mean"]
    var_rss23_median = this_df.describe()["var_rss23"]["50%"]
    var_rss23_std = this_df.describe()["var_rss23"]["std"]
    var_rss23_firstq = this_df.describe()["var_rss23"]["25%"]
    var_rss23_thirdq = this_df.describe()["var_rss23"]["75%"]
    var_rss23_all = [var_rss23_min, var_rss23_max, var_rss23_mean, var_rss23_median, var_rss23_std, var_rss23_firstq, var_rss23_thirdq]

    return [instance, test_or_train] + avg_rss12_all + var_rss12_all + avg_rss13_all + var_rss13_all + avg_rss23_all + var_rss23_all

for this_df in data_test:
    test_or_train = 0

    this_instance_features = get_time_features(this_df=this_df,test_or_train=test_or_train, instance=instance)

    time_domain_features.append(this_instance_features)

    instance += 1

for this_df in data_train:

    test_or_train = 1

    this_instance_features = get_time_features(this_df=this_df,test_or_train=test_or_train, instance=instance)

    time_domain_features.append(this_instance_features)

    instance += 1

column_names_avg12 = ["instance", "test_or_train", "avg12_min", "avg12_max", "avg12_mean", "avg12_median", "avg12_std", "avg12_firstq", "avg12_thirdq"]
column_names_var12 = ["var12_min", "var12_max", "var12_mean", "var12_median", "var12_std", "var12_firstq", "var12_thirdq"]
column_names_avg13 = ["avg13_min", "avg13_max", "avg13_mean", "avg13_median", "avg13_std", "avg13_firstq", "avg13_thirdq"]
column_names_var13 = ["var13_min", "var13_max", "var13_mean", "var13_median", "var13_std", "var13_firstq", "var13_thirdq"]
column_names_avg23 = ["avg23_min", "avg23_max", "avg23_mean", "avg23_median", "avg23_std", "avg23_firstq", "avg23_thirdq"]
column_names_var23 = ["var23_min", "var23_max", "var23_mean", "var23_median", "var23_std", "var23_firstq", "var23_thirdq"]

column_names = column_names_avg12 + column_names_var12 + column_names_avg13 + column_names_var13 + column_names_avg23 + column_names_var23

time_domain_df = pd.DataFrame(np.array(time_domain_features), columns=column_names)


#### iii. Standard Deviation

In [5]:
# print(time_domain_df.describe())
time_domain_stds_df = time_domain_df.describe().iloc[2]
print(time_domain_stds_df)

instance         25.547342
test_or_train     0.413809
avg12_min         9.569975
avg12_max         4.394362
avg12_mean        5.335718
avg12_median      5.440054
avg12_std         1.772153
avg12_firstq      6.153590
avg12_thirdq      5.138925
var12_min         0.000000
var12_max         5.062729
var12_mean        1.574164
var12_median      1.412244
var12_std         0.884105
var12_firstq      0.946386
var12_thirdq      2.125266
avg13_min         2.956462
avg13_max         4.875137
avg13_mean        4.008380
avg13_median      4.036396
avg13_std         0.946710
avg13_firstq      4.220658
avg13_thirdq      4.171628
var13_min         0.000000
var13_max         2.183625
var13_mean        1.166114
var13_median      1.145586
var13_std         0.458242
var13_firstq      0.843620
var13_thirdq      1.552504
avg23_min         6.124001
avg23_max         5.741238
avg23_mean        5.675593
avg23_median      5.813782
avg23_std         1.024898
avg23_firstq      6.096465
avg23_thirdq      5.531720
v

In [6]:
# Standard Deviations of the time-domain features:

# avg12_min         9.569975
# avg12_max         4.394362
# avg12_mean        5.335718
# avg12_median      5.440054
# avg12_std         1.772153
# avg12_firstq      6.153590
# avg12_thirdq      5.138925
# var12_min         0.000000
# var12_max         5.062729
# var12_mean        1.574164
# var12_median      1.412244
# var12_std         0.884105
# var12_firstq      0.946386
# var12_thirdq      2.125266
# avg13_min         2.956462
# avg13_max         4.875137
# avg13_mean        4.008380
# avg13_median      4.036396
# avg13_std         0.946710
# avg13_firstq      4.220658
# avg13_thirdq      4.171628
# var13_min         0.000000
# var13_max         2.183625
# var13_mean        1.166114
# var13_median      1.145586
# var13_std         0.458242
# var13_firstq      0.843620
# var13_thirdq      1.552504
# avg23_min         6.124001
# avg23_max         5.741238
# avg23_mean        5.675593
# avg23_median      5.813782
# avg23_std         1.024898
# avg23_firstq      6.096465
# avg23_thirdq      5.531720
# var23_min         0.045838
# var23_max         2.518921
# var23_mean        1.154812
# var23_median      1.086474
# var23_std         0.517617
# var23_firstq      0.758584
# var23_thirdq      1.523599

In [7]:
#[5]
time_domain_col_names = time_domain_df.columns.values.tolist()

time_domain_intervals = []

for this_col_name in time_domain_col_names:
    this_col_df = time_domain_df[this_col_name]

    this_col_array = np.array(this_col_df)
    #[6] [7]
    this_col_array = (this_col_array, )
    #[6] [7]
    this_col_results = bootstrap(this_col_array, np.std, confidence_level=0.9, method="percentile")

    time_domain_intervals.append([this_col_name, this_col_results.confidence_interval.low, this_col_results.confidence_interval.high])

time_domain_int_df = pd.DataFrame(time_domain_intervals, columns=["time_series_feature", "lower_bound", "upper_bound"])

print(time_domain_int_df)

   time_series_feature  lower_bound  upper_bound
0             instance    23.184488    27.217100
1        test_or_train     0.354829     0.450980
2            avg12_min     8.212807    10.750130
3            avg12_max     3.330210     5.238142
4           avg12_mean     4.681892     5.838307
5         avg12_median     4.764000     5.973226
6            avg12_std     1.558504     1.939846
7         avg12_firstq     5.534561     6.605173
8         avg12_thirdq     4.302042     5.814322
9            var12_min     0.000000     0.000000
10           var12_max     4.593852     5.378349
11          var12_mean     1.384899     1.696430
12        var12_median     1.228041     1.535231
13           var12_std     0.798626     0.936528
14        var12_firstq     0.827668     1.027429
15        var12_thirdq     1.888248     2.281940
16           avg13_min     2.742078     3.089455
17           avg13_max     4.149385     5.429403
18          avg13_mean     3.393033     4.468042
19        avg13_medi

In [8]:
#    time_series_feature  lower_bound  upper_bound
# 0             instance    23.221631    27.189456
# 1        test_or_train     0.354829     0.456247
# 2            avg12_min     8.234602    10.711499
# 3            avg12_max     3.325309     5.258153
# 4           avg12_mean     4.679485     5.855100
# 5         avg12_median     4.768167     5.961253
# 6            avg12_std     1.561170     1.939305
# 7         avg12_firstq     5.545619     6.614971
# 8         avg12_thirdq     4.307069     5.812099
# 9            var12_min     0.000000     0.000000
# 10           var12_max     4.598930     5.383356
# 11          var12_mean     1.383190     1.694530
# 12        var12_median     1.231027     1.535567
# 13           var12_std     0.798162     0.936493
# 14        var12_firstq     0.826688     1.029479
# 15        var12_thirdq     1.884852     2.283083
# 16           avg13_min     2.747719     3.094121
# 17           avg13_max     4.147518     5.437769
# 18          avg13_mean     3.403756     4.468950
# 19        avg13_median     3.412787     4.514471
# 20           avg13_std     0.758398     1.113562
# 21        avg13_firstq     3.597277     4.678926
# 22        avg13_thirdq     3.516211     4.651840
# 23           var13_min     0.000000     0.000000
# 24           var13_max     1.964227     2.347642
# 25          var13_mean     1.066697     1.216503
# 26        var13_median     1.048360     1.194443
# 27           var13_std     0.419664     0.483065
# 28        var13_firstq     0.769230     0.885296
# 29        var13_thirdq     1.425630     1.618564
# 30           avg23_min     4.393033     7.487510
# 31           avg23_max     4.722807     6.523554
# 32          avg23_mean     4.391391     6.690322
# 33        avg23_median     4.489899     6.862770
# 34           avg23_std     0.811236     1.213073
# 35        avg23_firstq     4.792746     7.173819
# 36        avg23_thirdq     4.356091     6.486624
# 37           var23_min     0.000000     0.078029
# 38           var23_max     2.239113     2.739993
# 39          var23_mean     1.057849     1.209059
# 40        var23_median     0.988929     1.141793
# 41           var23_std     0.476923     0.542358
# 42        var23_firstq     0.685978     0.802921
# 43        var23_thirdq     1.395187     1.591946

#### iv. Select Features

In selecting features, I'm going to try to look at the features that have the largest standard deviation. A feature with a large standard deviation means a high variance in values. This would imply that these values change the most across all the data.

To find this, I will calculate the average of the standard deviations for each measure (min, max, mean, etc.) across the six time series.

(Assistance from TA Zizhao Hu, who recommended that I take a look at the measures with the largest variance.)

In [9]:
min_average = (9.569975 + 0.000000 + 2.956462 + 0.000000 + 6.124001 + 0.045838)/6.0

max_average = (4.394362 + 5.062729 + 4.875137 + 2.183625 + 5.741238 + 2.518921)/6.0

mean_average = (5.335718 + 1.574164 + 4.008380 + 1.166114 + 5.675593 + 1.154812)/6.0

median_average = (5.440054 + 1.412244 + 4.036396 + 1.145586 + 5.813782 + 1.086474)/6.0

std_average = (1.772153 + 0.884105 + 0.946710 + 0.458242 + 1.024898 + 0.517617)/6.0

firstq_average = (6.153590 + 0.946386 + 4.220658 + 0.843620 + 6.096465 + 0.758584)/6.0

thirdq_average = (5.138925 + 2.125266 + 4.171628 + 1.552504 + 5.531720 + 1.523599)/6.0

print("min_average", min_average)
print("max_average", max_average)
print("mean_average", mean_average)
print("median_average", median_average)
print("std_average", std_average)
print("firstq_average", firstq_average)
print("thirdq_average", thirdq_average)

min_average 3.1160460000000003
max_average 4.129335333333333
mean_average 3.1524634999999996
median_average 3.155756
std_average 0.9339541666666668
firstq_average 3.1698838333333335
thirdq_average 3.3406070000000003


Results:

min_average 3.1160460000000003

max_average 4.129335333333333

mean_average 3.1524634999999996

median_average 3.155756

std_average 0.9339541666666668

firstq_average 3.1698838333333335

thirdq_average 3.3406070000000003

So it appears that the measures that have the highest standard deviations across the six time series are "max", "third quartile", and "first quartile". Thus, I will be using those measures.

## 2. ISLR 3.7.4

### (a) Linear Train

I would say that there is not enough information.

A cubic function can be seen as linear at a small enough range. (For instance, y = x^3 looks very linear very close to x=0.) Thus, even if the underlying true relationship is linear, the cubic function can be fitted and scaled such that it is very close to linear around the range of the training data. Due to noise, this could cause the RSS for the cubic regression model to actually be lower than the RSS of the linear regression model. (It could just so happen that the cubic regression model fits this certain training dataset better.)

In addition, a cubic regression model could also overfit the data. This will reduce the cubic model's RSS for the training data, and could make it less than the linear model's RSS.

All of this doesn't mean that the linear regression model will always have a greater RSS than the cubic model -- it could have a smaller RSS than the cubic model. However, given that we don't know how much noise is in the data, I would say that there isn't enough information to make a strong conclusion either way.

### (b) Linear Test

I would say that the linear regression model will have a lower RSS.

As the true underlying relationship is linear, it's much more likely that the new test data will generally fit more closely the linear regression model, as opposed to the cubic regression model. This is because outside of the range of the training data, the cubic regression model could greatly diverge from the underlying linear relationship. (y = x^3 is close to linear for a small section near x=0, but farther away from that, it widely diverges from being linear.) In addition, the cubic regression model could be overfitted, meaning that it will not be able to generalize that well to new data as compared to the linear regression model.

### (c) Not Linear Train

I would say we do not have enough information.

Like part a, we don't know how much noise is in the data. Furthermore, a cubic function can be fitted and scaled to be linear-like (or even quadratic-like) around the range of the test data, possibly resulting in a lower RSS for the cubic function as compared to the linear function. (For instance, if the underlying relationship is quadratic, the cubic function can be scaled/fitted to be quadratic-like around the range of the training data. This would then perform better than the linear regression model.)

Most importantly, however, is the fact that we do not know how far away from linear our true underlying relationship is. If our true underlying relationship is very close to linear, then the linear regression model will tend to have a lower RSS value. However, if the true underlying relationship is much farther away from linear, then the linear regression model will tend to have a higher RSS value.

Given all these uncertainties, I would say we do not have enough information to make a reasonable expectation.

### (d) Not Linear Testing

I would say we do not have enough information.

As mentioned in part c, we do not know how close or far away the true underlying relationship is from being linear. If the true relationship is close to linear, then it's more likely that the linear regression model will fit the new test data well, and less likely that the cubic regression model will fit the test data well, resulting in a lower RSS for the linear regression model as compared to the cubic regression model. On the other hand, the true relationship could be closer to cubic, meaning that the cubic model could fit the true relationship much better than the linear model, meaning that the cubic model will tend to have a much lower RSS as compared to the cubic model.

In other words, given that we don't know the underlying true relationship, it's difficult to judge if a linear or a cubic model will work better.

## 3. ISLR 3.7.3 - Extra Practice 

## 3. ISLR 3.7.5 - Extra Practice 

References:

[1] https://www.geeksforgeeks.org/how-to-iterate-over-files-in-directory-using-python/#

[2] https://www.w3schools.com/python/pandas/pandas_csv.asp

[3] https://towardsdatascience.com/a-brief-introduction-to-time-series-classification-algorithms-7b4284d31b97

[4] https://machinelearningmastery.com/machine-learning-data-transforms-for-time-series-forecasting/

[5] https://sparkbyexamples.com/pandas/pandas-get-column-names/

[6] https://www.statology.org/bootstrapping-in-python/

[7] https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bootstrap.html